presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -89,400 +111,733 @@
+                node.getName(),
+                processSameType(node.getQuery()),
+                node.getColumnNames());
     }
 
-    @Override
-    protected Node visitOrderBy(OrderBy node, Void context)
+    private <T extends Node> List<T> processNodes(List<T> nodes)
     {
-        ImmutableList.Builder<SortItem> rewrittenOrderBy = ImmutableList.builder();
-        for (SortItem sortItem : node.getSortItems()) {
-            sortItem = removeSortItemPrefix(sortItem, removablePrefix);
-            if (!materializedViewInfo.getBaseToViewColumnMap().containsKey(sortItem.getSortKey())) {
-                throw new IllegalStateException(format("Sort key %s is not present in materialized view select fields", sortItem.getSortKey()));
+        List<T> newNodes = new ArrayList<>();
+        boolean listsIdentical = true;
+
+        for (T node : nodes) {
+            T newNode = processSameType(node);
+            if (node != newNode) {
+                listsIdentical = false;
             }
-            rewrittenOrderBy.add((SortItem) process(sortItem, context));
+            newNodes.add(newNode);
         }
-        return new OrderBy(rewrittenOrderBy.build());
+        return listsIdentical ? nodes : newNodes;
     }
 
-    @Override
-    protected Node visitSortItem(SortItem node, Void context)
+    private <T extends Node> T processSameType(T node)
     {
-        return new SortItem((Expression) process(node.getSortKey(), context), node.getOrdering(), node.getNullOrdering());
+        return (T) process(node);
     }
 
-    @Override
-    protected Node visitSimpleGroupBy(SimpleGroupBy node, Void context)
+    private QuerySpecification rewriteQuerySpecificationIfCompatible(QuerySpecification querySpecification, Table baseTable)
     {
-        ImmutableList.Builder<Expression> rewrittenSimpleGroupBy = ImmutableList.builder();
-        for (Expression column : node.getExpressions()) {
-            rewrittenSimpleGroupBy.add((Expression) process(removeExpressionPrefix(column, removablePrefix), context));
+        QualifiedObjectName tableName = createQualifiedObjectName(session, baseTable, baseTable.getName());
+        List<ConnectorMaterializedViewDefinition> linkedMaterializedViews = baseTableToMaterializedViewMap.getOrDefault(tableName, ImmutableList.of());
+
+        for (ConnectorMaterializedViewDefinition materializedViewDefinition : linkedMaterializedViews) {
+            Table materializedViewTable = new Table(QualifiedName.of(materializedViewDefinition.getTable()));
+            Query materializedViewQuery = (Query) sqlParser.createStatement(materializedViewDefinition.getOriginalSql(), createParsingOptions(session));
+
+            QuerySpecification rewritten = new QuerySpecificationRewriter(materializedViewTable, materializedViewQuery).rewrite(querySpecification);
+
+            if (rewritten != querySpecification) {
+                session.getRuntimeStats().addMetricValue(OPTIMIZED_WITH_MATERIALIZED_VIEW_COUNT, NONE, 1);
```

Comment:
I think we should create new run time metric for subqueries. The idea is to be able to monitor optimized subquery and query count independently.

Reply:
This new metric would be a superset of the existing one, correct? We increment the existing metric once for a given query if any optimization occurs, and increment the new metric for each subquery optimization

Reply:
I have a WIP PR open related to metrics, here's the relevant commit: https://github.com/prestodb/presto/pull/17949/commits/e339a5baa40931b1133babbcb4f649aa48cfec53. It might be good to add the new metric there?